Issue #215: Do not contribute SCB layer if already there

anthonydahanne opened the pull request:
* prefer the spring-cloud-bindings jar provided by the user

## Summary
Add a check on the presence of an existing spring-cloud-bindings jar in the Spring libs (BOOT-INF/libs)

## Use Cases
When a user already added to its application a spring-cloud-bindings version, the spring-boot buildpack should not add another one.

## Checklist

* [X] I have viewed, signed, and submitted the Contributor License Agreement.
* [X] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [X] I have added an integration test, if necessary.
* [X] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [X] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

pivotal-david-osullivan commented on boot/build.go:
```diff
@@ -131,8 +131,12 @@ func (b Build) Build(context libcnb.BuildContext) (libcnb.BuildResult, error) {
 	} else {
+		scbJar := FindSpringCloudBindingsJars(d)
+		if scbJar != nil {
+			b.Logger.Headerf("A Spring Cloud Bindings library was found in the Spring Boot libs with this version: %s - not adding another one", scbJar.Version)
```
Do you think it important that the version is given in the output here? I'm thinking that it might be enough to let the user know the buildpack found the lib and that it will not contribute one.

anthonydahanne replied:
no, not really. Let's keep things simple and focused

pivotal-david-osullivan commented on boot/build.go:
```diff
@@ -324,3 +328,14 @@ func friendlySize(size float64) string {
 	return fmt.Sprintf("%0.1f %s", size, unit)
 }
+
+func FindSpringCloudBindingsJars(jars []libjvm.MavenJAR) *libjvm.MavenJAR {
```
You could consider making this func generic, e.g. FindExistingDependency and pass in the name you want to match.

If you don't think the version is needed you could simplify this func and simply return a bool

anthonydahanne replied:
yes, good idea

pivotal-david-osullivan commented on boot/build.go:
```diff
@@ -324,3 +328,14 @@ func friendlySize(size float64) string {
+	springCloudBindingsName := "spring-cloud-bindings"
+	for _, lib := range jars {
+		if lib.Name == springCloudBindingsName {
+			springCloudBindingsJar = &lib
```
There is a bug here. The version output in my log is not correct, I was getting '9.0.69' which is the Tomcat lib versions. This is because you are assigning springCloudBindingsJar the address of 'lib' which changes on each loop, thereby it always ends up with the last value after the first time the 'if' statement returns true.

If you **do** keep the version extraction logic and thus keep returning the reference, you could break/return early here, there's no need to continue the loop once you find a match anyway. You could also simply pass the value instead of using pointers.

dmikusa replied:
I would add a follow-up point here, in that this bug slipped through the unit tests. The test coverage looks good on the surface, but something must be off since this didn't get caught. My suggestion would be to modify the unit tests to catch the bug that @pivotal-david-osullivan pointed out and then fix it.

anthonydahanne replied:
I could reproduce the issue in the test, after appending another `libjvm.MavenJAR` into `mavenJars`

I missed that bug because in my test the `springCloudBindingsJar` was the last jar I added.

That said, I'll probably go down the route suggested by @pivotal-david-osullivan , ie not returning the version anymore, and stopping the loop after the first hit, returning a boolean

anthonydahanne commented:
hi there @dmikusa @pivotal-david-osullivan !
I try re requesting your reviews via GH UI, but apparently only one can be set as a reviewer!
Anyway, please feel free to re review; I should have addressed all concerns with my updated commit

pivotal-david-osullivan commented on boot/build.go:
```diff
@@ -131,8 +131,12 @@ func (b Build) Build(context libcnb.BuildContext) (libcnb.BuildResult, error) {
 
 	} else {
+		scbJarFound := FindExistingDependency(d, "spring-cloud-bindings")
+		if scbJarFound == true {
```
```suggestion
		if scbJarFound {
```
Small simplification!

pivotal-david-osullivan commented on boot/build.go:
```diff
@@ -131,8 +131,12 @@ func (b Build) Build(context libcnb.BuildContext) (libcnb.BuildResult, error) {
+		}
 		// contribute Spring Cloud Bindings - false by default
-		if !cr.ResolveBool("BP_SPRING_CLOUD_BINDINGS_DISABLED") {
+		if !cr.ResolveBool("BP_SPRING_CLOUD_BINDINGS_DISABLED") && scbJarFound == false {
```
```suggestion
		if !cr.ResolveBool("BP_SPRING_CLOUD_BINDINGS_DISABLED") && !scbJarFound {
```

anthonydahanne commented:
Can you please add the following labels please ?  `semver:patch` and `type:enhancement` (or `type:bug` even?!)
thank you!
